### ISSUE TITLE
feat: apisix support QUIC+HTTP3

### ISSUE THREAD

**Issue by lizj3624** [apache/apisix#6403]
### Issue description

[cloudflare/quiche](https://github.com/cloudflare/quiche) patch nginx-1.16 support quic+http3, I hope 
apisix support QUIC+HTTP3

**Comment by moonming** [comment]
great idea! I'm not familiar with quiche, how much would it cost to use it instead of openresty? In addition, nginx is officially merging the functions of QUIC+HTTP3?
cc @membphis

**Comment by lizj3624** [comment]
I patch [quiche](https://github.com/cloudflare/quiche) to OpenResty-1.15.8，I found error (epoll_ctl(1, 53) failed (17: File exists)) when access_by_lua of ngx_lua module enabled，I gruess ngx_lua not support quic.  [ngx_lua issue](https://github.com/openresty/lua-nginx-module/issues/1887)

1. build env
```shell
nginx version: openresty/1.15.8.1 (quiche-09c0e37)
built by gcc 4.8.5 20150623 (Red Hat 4.8.5-36) (GCC)
built with OpenSSL 1.1.1 (compatible; BoringSSL) (running with BoringSSL)
TLS SNI support enabled
configure arguments: --prefix=/myngx/nginx --with-debug --with-cc-opt='-DNGX_LUA_USE_ASSERT -DNGX_LUA_ABORT_AT_PANIC -O2 -ggdb -O2 -D JD_NGX_SSL_HANDSHAKE_TIME -D JD_NGX_HTTP_UPSTREAM_RANDOM' --add-module=../ngx_devel_kit-0.3.1rc1 --add-module=../echo-nginx-module-0.61 --add-module=../xss-nginx-module-0.06 --add-module=../ngx_coolkit-0.2 --add-module=../set-misc-nginx-module-0.32 --add-module=../form-input-nginx-module-0.12 --add-module=../encrypted-session-nginx-module-0.08 --add-module=../srcache-nginx-module-0.31 --add-module=../ngx_lua-0.10.15 --add-module=../ngx_lua_upstream-0.07 --add-module=../headers-more-nginx-module-0.33 --add-module=../array-var-nginx-module-0.05 --add-module=../memc-nginx-module-0.19 --add-module=../redis2-nginx-module-0.15 --add-module=../redis-nginx-module-0.3.7 --add-module=../rds-json-nginx-module-0.15 --add-module=../rds-csv-nginx-module-0.09 --add-module=../ngx_stream_lua-0.0.7 --with-ld-opt=-Wl,-rpath,/export/servers/myngx/luajit/lib --build=quiche-09c0e37 --with-pcre=/root/rpmbuild/BUILD/myngx-2.3-52.d375389/thirdparty/pcre-8.39 --with-pcre-jit --with-zlib=/root/rpmbuild/BUILD/myngx-2.3-52.d375389/thirdparty/zlib-1.2.8 --with-threads --with-http_auth_request_module --with-http_ssl_module --with-http_gzip_static_module --with-http_stub_status_module --with-http_v2_module --with-http_realip_module --with-http_addition_module --with-http_slice_module --add-module=/root/rpmbuild/BUILD/myngx-2.3-52.d375389/thirdparty/lua-ssl-nginx-module --add-module=/root/rpmbuild/BUILD/myngx-2.3-52.d375389/thirdparty/ngx_http_dyups_module --add-module=/root/rpmbuild/BUILD/myngx-2.3-52.d375389/thirdparty/ngx_http_sticky_module --with-stream --with-stream_ssl_module --with-http_v3_module --with-openssl=/root/myhome/quiche-cloudflare/deps/boringssl --with-quiche=/root/myhome/quiche-cloudflare --with-stream --with-stream_ssl_preread_module
``` 

2.  some code
ngx的conf配置
```cond
http {
...
lua_max_running_timers 512;
lua_max_pending_timers 1024;

lua_package_path "$prefix/app/lib/?.lua;$prefix/app/etc/?.lua;$prefix/app/src/?.lua;$prefix/sbin/openresty/lualib/?.lua;;";
lua_package_cpath "$prefix/app/lib/?.so;$prefix/sbin/openresty/lualib/?.so;;";

lua_check_client_abort on;
lua_socket_log_errors off;
lua_http10_buffering off;
...

server {
listen :443 quic reuseport;
isten :443 ssl http2;
...
access_by_lua_file my_waf.lua
...

}
}
```
WAF模块的相关代码逻辑
my_waf.lua
```lua
_M.waf_request = function ()
...ngx.req.get_body_file()..
local http_conn = http.new()
...http_conn.connect()...
...http_conn.request()...
...

end
```

3. error msg
```conf
2022/02/22 11:00:05 [alert] 27764#27764: *42474256 epoll_ctl(1, 36) failed (17: File exists), client: 1.114.102.56, server:quic.ngx.com, request: "POST /a?&t=1511418329196 HTTP/2.0", host: "quic.ngx.com"
```

**Comment by spacewander** [comment]
We have discussed it before. It seems that Nginx has developed its HTTP3 support instead of quiche. Since we need to catch up with the latest Nginx, it would be better to use Nginx's own HTTP3 implementation.

https://github.com/cloudflare/quiche/tree/master/nginx#building
We can't go back to Nginx 1.16 to use quiche

**Comment by lizj3624** [comment]
@spacewander when apisix support QUIC+HTTP3？

**Comment by spacewander** [comment]
Once this feature is merged in Nginx and we catch up the latest version.

**Comment by lizj3624** [comment]
> You can make a try: https://hg.nginx.org/nginx-quic/shortlog/quic + openresty @lizj3624

@membphis thx，from the test results，[quiche's](https://github.com/cloudflare/quiche) stability is better than nginx-quic, cloudflare‘s quiche has been running online for a long time，nginx-quic has not been merged into the master branch。

**Comment by spacewander** [comment]
Yes, but we are not Cloudflare. The Cloudflare-Nginx is a fork from an old Nginx version. We don't want to make APISIX hard to catch up with the latest Nginx.

**Comment by lizj3624** [comment]
1.  OK，I understand apisix
2. I complain about nginx，official nginx plans merge quic to master before 2022，[nginx h3 roadmap](https://www.nginx.com/blog/our-roadmap-quic-http-3-support-nginx/)，but now is 2022.2，nginx new feat is too slow，alibaba，tencent，kuaishou  LB supports quic，We are also testing cloudflare-quic nginx
3. Igor Sysoev，nginx's father，resigned from nginx.Inc and F5 on 2022.1.18

**Comment by gacktfly** [comment]
any plans for this？

**Comment by moonming** [comment]
> any plans for this？

I think APISIX can be upgraded with the new version of NGINX

**Comment by gacktfly** [comment]
> > any plans for this？
> 
> I think APISIX can be upgraded with the new version of NGINX

great，cant wait to see it

**Comment by monkeyDluffy6017** [comment]
Please refer to this issue: https://github.com/apache/apisix/issues/9752

**Comment by NiuBlibing** [comment]
Hi, [openresty 1.25.3.1](https://github.com/openresty/openresty/releases/tag/v1.25.3.1) has been released, is it ready to upgrade it for apisix?

**Comment by monkeyDluffy6017** [comment]
Yeah, we will upgrade ASAP

**Comment by monkeyDluffy6017** [comment]
@NiuBlibing I hope so

**Comment by zll600** [comment]
@monkeyDluffy6017 I want to do this, please assign to me

**Comment by SilentEntity** [comment]
@zll600, let's work together on this one. [co-work]

**Comment by zll600** [comment]
@SilentEntity Thank you. Here is my plan. I want to upgrade openresty-1.25.3.1 to make apisix support QUIC+HTTP3. I will be focusing on doing this for the next few days, if you have any ideas please let me know.

#### Step one
- [ ] upgrade patch file in  https://github.com/api7/ngx_multi_upstream_module
- [ ] upgrade patch file in https://github.com/api7/apisix-nginx-module
- [ ] Upgrade apisix-runtime to use openresty-1.25.3.1 in https://github.com/api7/apisix-build-tools
- [ ] Upgrade apisix to use new version of apisix-runtime

#### Step two
- [ ] Provide quic configuration in apisix so that basic plug-ins can use quic

#### Step Three
- [ ] Test quic’s compatibility under all plugins.

**Comment by NiuBlibing** [comment]
Congratulations! May I ask how soon the new version will be released?
